src/symbols.c

```diff
@@ -0,0 +1,85 @@
+/*
+ * SPDX-License-Identifier: BSD-2-Clause
+ *
+ * Copyright 2010-2022, Tarantool AUTHORS, please see AUTHORS file.
+ */
+#include <string.h>
+
+/*
+ * This file is supposed to be autogenerated in a future.
+ *
+ * Now it contains several internal symbols we want to offer
+ * outside of the module API: no guarantees are provided
+ * regarding them.
+ *
+ * fiber_channel_*() and ipc_value_*() symbols are used in
+ * the Rust module [1], because the symbols were exported in
+ * Tarantool 2.8.
+ *
+ * [1]: https://github.com/picodata/tarantool-module
+ */
+
+extern void
+fiber_channel_close(void);
+extern void
+fiber_channel_create(void);
+extern void
+fiber_channel_delete(void);
+extern void
+fiber_channel_destroy(void);
+extern void
+fiber_channel_get_msg_timeout(void);
+extern void
+fiber_channel_get_timeout(void);
+extern void
+fiber_channel_has_readers(void);
+extern void
+fiber_channel_has_writers(void);
+extern void
+fiber_channel_new(void);
+extern void
+fiber_channel_put_msg_timeout(void);
+extern void
+fiber_channel_put_timeout(void);
+extern void
+ipc_value_delete(void);
+extern void
+ipc_value_new(void);
+
+/** Symbol definition. */
+struct symbol_def {
+	/** Name of the symbol. */
+	const char *name;
+	/** Address of the symbol. */
+	void *addr;
+};
+
+static struct symbol_def symbols[] = {
```

Comment:
While the solution looks fine (except I would rename the function to `tnt_dlsym` or `tnt_sym`), I more like Timur's way: https://github.com/tarantool/tarantool/pull/7414. It allows to add symbol aliases with `tnt_` or any other prefix quite easy and users won't need to bother about finding the addresses at runtime. Moreover it would inevitable force them in any not too small project to introduce global variables which would hold the pointers. It reminds me of how hard it is to use non-blocking accept in Windows API - you need to get accept-function pointer per-socket and drag it everywhere with the socket.

What was wrong with Timur's PR? Or did you not see it yet?

Reply:
Those are different directions. Timur refactors machinery of exposing public symbols (with manual marking of the symbols). Here I perform a step toward automatic exposing of internal symbols without spoiling of the global namespace (see also PR #7424). I would propose the automation without this transient state with hardcoded `fiber_channel_*` list, but I need some considerable time to make this machinery stable enough.

I would prefer to don't spoil the public symbols namespace with internal/temporary symbols, even with a prefix. It just a bit, well, dirty. Also I don't like flips forth and back in the git history, when we marking exposed internal symbols manually. And the manually marked internal symbols will only available in new releases.

If we'll proceed with something like PR #7424, it should just work. We can even clean up symbols which are exported purely for internal usage via FFI and get rid of some `tnt_*()` wrappers.

As result, I hope, I'll have the same set of declarations in `module.h` and in the export list (disregarding whether it is represented as the linker script or the tricky macro). Or will be closer to this state at least. Maybe even will have some checkers/linters around.

----

Regarding Timur's macro per se: I don't know, whether it is portable and stable enough. However I don't mind the idea in general. Need to think around.

----

Regarding name. I like the short word `sym`, I spinned around it initially. However:

1. `dl` (dynamic loader) is irrelevant here, so not `dlsym`.
2. I want `internal` in the name, because it is very dangerous function.
3. `internal` is wrote as the full word, so `sym` would look strange. In those circumstances `symbol` looks better.
4. Maybe it should look long and ugly, because nobody should guess that it is something safe to use.

If you have something short, but visually dangerous, I'm open for proposals.

Reply:
> Those are different directions.

In this PR I see that you list the exported symbols manually just like Timur does. In that sense I don't see any pros against Timur's way. His approach allows to export the symbols with `tnt_internal_` prefix if you want the user to see "internal" keyword and it still allows the user to remember the function address into a variable with a smaller name.

But if we have #7424, then why is this PR needed?

> disregarding whether it is represented as the linker script or the tricky macro

From what I see, the only win of #7424 against Timur asm hacks is that the script allows not to mark each symbol individually. If I understood you correctly as the tricky macro you mean marking each symbol manually? Then how is it better than Timur's PR? Or did I misunderstand you?

> If you have something short, but visually dangerous, I'm open for proposals.

`tnt_sym()` seemed dangerous enough to me. As for `_dlsym` suffix - it works similarly to `dlsym()` in a sense of observable behaviour, so such suffix would look easy to catch what the function is doing. As for contractions - I don't like long names regardless of their "internalness". If a guy has decided to use internal symbols, the word "internal" is unlikely to stop him. The fact of not even being able to see the function signature is a big obstacle enough IMO. But I won't argue here further, any name works for me.

Reply:
I have not much to add to said above, but we discussed it verbally and, it seems, Vlad agreed with the arguments above.

To sum up:

* It is intermediate state toward #7424. We need *this* PR to polish #7424 during some time, but at the same time offer the fiber channel symbols in the upcoming 2.10.1 release.
* It has no relation to `extra/exports` vs macro with `asm()` exporting approach, because here we don't export symbols into the executable symbols directly.

----

Regarding names: we both stuck on our opinions regarding the best name, but Vlad don't want to insist on his one.